xdsl/irdl/irdl.py

```diff
@@ -381,6 +381,40 @@ def get_unique_base(self) -> type[Attribute] | None:
         return None
 
 
+@dataclass(init=False)
+class MessageConstraint(AttrConstraint):
+    """
+    Attach a message to a constraint, to provide more context when the constraint
+    is not satisfied.
+    """
+
+    constr: AttrConstraint
+    message: str
+
+    def __init__(
+        self, constr: AttrConstraint | Attribute | type[Attribute], message: str
+    ):
+        if isinstance(constr, type):
+            constr = BaseAttr(constr)
+        elif isinstance(constr, Attribute):
+            constr = EqAttrConstraint(constr)
+
+        self.constr = constr
+        self.message = message
+
+    def verify(self, attr: Attribute, constraint_vars: dict[str, Attribute]) -> None:
+        try:
+            return self.constr.verify(attr, constraint_vars)
+        except VerifyException as e:
+            raise VerifyException(e.args[0] + f"\n{self.message}", *e.args[1:])
```

Comment:
Can you please add a comment explaining the indexing logic here?

Reply:
Maybe add a check for e.args[0] always being present?

Reply:
Agreed on this code looking cryptic, OTOH this is simply the generic Exception. Could we maybe make our exception class itself more readable? :thinking:

Reply:
I'll do it as a bit of a meditative exercise, I think the PR is ok to merge as-is